### PR TITLE
Test full timestamp output range in PySpark [databricks]

### LINKED ISSUE #9747: [FEA] Figure out the right long term way to deal with dates and timestamps in integration tests.
**Is your feature request related to a problem? Please describe.**
Python/Pyspark do not really support dates/timestamps that are prior to the year 1.  This can cause some issues in some cases as Spark will happily produce values that cannot be shipped back to python. Even before https://github.com/NVIDIA/spark-rapids/pull/9736 the current code restricted the range of values that could be generated to try and avoid this issue. But there is nothing that prevents a test from subtracting a month, or year, or whatever from a timestamp that would result in a value being out of range.

I think it would be better to have timestamp and data gen produce a full range of possible supported values. Especially for corner case tests with timezones/etc. We could also have the code automatically convert dates and timestamps to the raw number representation using `date_diff` and `unix_micros`. But there was also the suggestion to try and do both. But I am not 100% sure how to do both of them and coordinate things. So this is here to try and figure out if it is possible or if there is a better solution we have not thought of yet.

### PR DESCRIPTION
**Pull request by NVnavkumar**
Fixes #9747.

This handles timestamp values from Spark that are out of range in Python.  We use the `patch` facility in `unittest` from standard Python to override the `fromInternal` implementation of the pyspark TimestampType, TimestampNTZType, and DateType. This allows us to compare the microsecond output of both the CPU and GPU without processing to a python `datetime` or `date` object which can be subject to rounding errors.

### CODE REVIEW

File: integration_tests/src/main/python/conftest.py
Comment (revans2) [inline integration_tests/src/main/python/conftest.py:42]: Sadly TimestampNTZ was added to spark/pyspark in 3.3.0 https://issues.apache.org/jira/browse/SPARK-37126 and we still support some older versions of Spark. So we need to make this work for all versions we support.

**Review by revans2 — COMMENTED** [review]
Looks good to me. I would like to see some more documentation about what it is doing, and I don't like the name `monkey_patches`. If we could rename it to something more descriptive that would be great.

File: integration_tests/src/main/python/data_gen.py
Comment (revans2) [inline integration_tests/src/main/python/data_gen.py:583]: nit: Could you file a follow on issue to look at how we might be able to work around the range limitations on input to timestamps and dates? It might be nice to inject some processing or something to make it so that we can do the full range that spark supports instead of just what python can handle.

**Review by revans2 — APPROVED** [review]

**Comment by NVnavkumar** [comment]
build

File: integration_tests/src/main/python/data_gen.py
Comment (NVnavkumar) [reply integration_tests/src/main/python/data_gen.py:583]: So to clarify, would this mean supporting negative timestamps that Spark can handle since the underlying value is microseconds? Or is there something missing here from the `0001-01-01` to `9999-12-31` range?

**Comment by NVnavkumar** [comment]
build

File: integration_tests/src/main/python/spark_session.py
Comment (NVnavkumar) [inline integration_tests/src/main/python/spark_session.py:91]: Will file a follow up issue to handle diff usability.

File: integration_tests/src/main/python/data_gen.py
Comment (revans2) [reply integration_tests/src/main/python/data_gen.py:583]: Yes it would be nice to handle generating negative timestamp values, and positive values that support the full range that Spark supports. Up to 6 digit years. I think this is minor, but it would nice to have an issue to look into it.

File: integration_tests/src/main/python/spark_session.py
Comment (NVnavkumar) [reply integration_tests/src/main/python/spark_session.py:91]: https://github.com/NVIDIA/spark-rapids/issues/10039

File: integration_tests/src/main/python/data_gen.py
Comment (NVnavkumar) [reply integration_tests/src/main/python/data_gen.py:583]: https://github.com/NVIDIA/spark-rapids/issues/10040

File: integration_tests/src/main/python/data_gen.py
Comment (NVnavkumar) [reply integration_tests/src/main/python/data_gen.py]: I updated this to adding `start` and `end` as special cases. But this is important (especially in the `min_start` case) since that case triggers the `value 0 out of range error` that we have been seeing in earlier failures.

**Comment by NVnavkumar** [comment]
From premerge CI, looks like there is some place where this patch is not applying (looks like in hash_aggregate_test), will take a look.

**Comment by NVnavkumar** [comment]
Hmm, it looks like the failure is in PythonRunner (so pyspark code running inside an executor). Currently this patch code can patch driver code, not executor code.

```
E                   py4j.protocol.Py4JJavaError: An error occurred while calling o677.collectToPython.
E                   : org.apache.spark.SparkException: Job aborted due to stage failure: Task 2 in stage 5.0 failed 1 times, most recent failure: Lost task 2.0 in stage 5.0 (TID 22) (192.168.4.31 executor driver): org.apache.spark.api.python.PythonException: Traceback (most recent call last):
E                     File "/home/navink/spark/spark-3.1.1-bin-hadoop3.2/python/lib/pyspark.zip/pyspark/worker.py", line 604, in main
E                       process()
E                     File "/home/navink/spark/spark-3.1.1-bin-hadoop3.2/python/lib/pyspark.zip/pyspark/worker.py", line 596, in process
E                       serializer.dump_stream(out_iter, outfile)
E                     File "/home/navink/spark/spark-3.1.1-bin-hadoop3.2/python/lib/pyspark.zip/pyspark/serializers.py", line 259, in dump_stream
E                       vs = list(itertools.islice(iterator, batch))
E                     File "/home/navink/spark/spark-3.1.1-bin-hadoop3.2/python/lib/pyspark.zip/pyspark/serializers.py", line 138, in load_stream
E                       yield self._read_with_length(stream)
E                     File "/home/navink/spark/spark-3.1.1-bin-hadoop3.2/python/lib/pyspark.zip/pyspark/serializers.py", line 160, in _read_with_length
E                       return self.loads(obj)
E                     File "/home/navink/spark/spark-3.1.1-bin-hadoop3.2/python/lib/pyspark.zip/pyspark/serializers.py", line 430, in loads
E                       return pickle.loads(obj, encoding=encoding)
E                     File "/home/navink/spark/spark-3.1.1-bin-hadoop3.2/python/lib/pyspark.zip/pyspark/sql/types.py", line 1416, in <lambda>
E                       return lambda *a: dataType.fromInternal(a)
E                     File "/home/navink/spark/spark-3.1.1-bin-hadoop3.2/python/lib/pyspark.zip/pyspark/sql/types.py", line 651, in fromInternal
E                       values = [f.fromInternal(v) if c else v
E                     File "/home/navink/spark/spark-3.1.1-bin-hadoop3.2/python/lib/pyspark.zip/pyspark/sql/types.py", line 651, in <listcomp>
E                       values = [f.fromInternal(v) if c else v
E                     File "/home/navink/spark/spark-3.1.1-bin-hadoop3.2/python/lib/pyspark.zip/pyspark/sql/types.py", line 452, in fromInternal
E                       return self.dataType.fromInternal(obj)
E                     File "/home/navink/spark/spark-3.1.1-bin-hadoop3.2/python/lib/pyspark.zip/pyspark/sql/types.py", line 319, in fromInternal
E                       return obj and [self.elementType.fromInternal(v) for v in obj]
E                     File "/home/navink/spark/spark-3.1.1-bin-hadoop3.2/python/lib/pyspark.zip/pyspark/sql/types.py", line 319, in <listcomp>
E                       return obj and [self.elementType.fromInternal(v) for v in obj]
E                     File "/home/navink/spark/spark-3.1.1-bin-hadoop3.2/python/lib/pyspark.zip/pyspark/sql/types.py", line 319, in fromInternal
E                       return obj and [self.elementType.fromInternal(v) for v in obj]
E                     File "/home/navink/spark/spark-3.1.1-bin-hadoop3.2/python/lib/pyspark.zip/pyspark/sql/types.py", line 319, in <listcomp>
E                       return obj and [self.elementType.fromInternal(v) for v in obj]
E                     File "/home/navink/spark/spark-3.1.1-bin-hadoop3.2/python/lib/pyspark.zip/pyspark/sql/types.py", line 319, in fromInternal
E                       return obj and [self.elementType.fromInternal(v) for v in obj]
E                     File "/home/navink/spark/spark-3.1.1-bin-hadoop3.2/python/lib/pyspark.zip/pyspark/sql/types.py", line 319, in <listcomp>
E                       return obj and [self.elementType.fromInternal(v) for v in obj]
E                     File "/home/navink/spark/spark-3.1.1-bin-hadoop3.2/python/lib/pyspark.zip/pyspark/sql/types.py", line 186, in fromInternal
E                       return datetime.datetime.fromtimestamp(ts // 1000000).replace(microsecond=ts % 1000000)
E                   ValueError: year 0 is out of range
E
E                   	at org.apache.spark.api.python.BasePythonRunner$ReaderIterator.handlePythonException(PythonRunner.scala:517)
E                   	at org.apache.spark.api.python.PythonRunner$$anon$3.read(PythonRunner.scala:652)
E                   	at org.apache.spark.api.python.PythonRunner$$anon$3.read(PythonRunner.scala:635)
E                   	at org.apache.spark.api.python.BasePythonRunner$ReaderIterator.hasNext(PythonRunner.scala:470)
E                   	at org.apache.spark.InterruptibleIterator.hasNext(InterruptibleIterator.scala:37)
E                   	at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:489)
E                   	at scala.collection.Iterator$$anon$10.hasNext(Iterator.scala:458)
E                   	at scala.collection.Iterator$$anon$10.hasNext(Iterator.scala:458)
E                   	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIteratorForCodegenStage1.processNext(Unknown Source)
E                   	at org.apache.spark.sql.execution.BufferedRowIterator.hasNext(BufferedRowIterator.java:43)
E                   	at org.apache.spark.sql.execution.WholeStageCodegenExec$$anon$1.hasNext(WholeStageCodegenExec.scala:755)
E                   	at org.apache.spark.sql.execution.SparkPlan.$anonfun$getByteArrayRdd$1(SparkPlan.scala:345)
E                   	at org.apache.spark.rdd.RDD.$anonfun$mapPartitionsInternal$2(RDD.scala:898)
E                   	at org.apache.spark.rdd.RDD.$anonfun$mapPartitionsInternal$2$adapted(RDD.scala:898)
E                   	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
E                   	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:373)
E                   	at org.apache.spark.rdd.RDD.iterator(RDD.scala:337)
E                   	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)
E                   	at org.apache.spark.scheduler.Task.run(Task.scala:131)
E                   	at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$3(Executor.scala:497)
E                   	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1439)
E                   	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:500)
E                   	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
E                   	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
E                   	at java.lang.Thread.run(Thread.java:750)
```

**Comment by NVnavkumar** [comment]
I think the issue is in the use of `createDataFrame` here:

```
def test_hash_groupby_collect_set_on_nested_array_type(data_gen):
    conf = copy_and_update(_float_conf, {
        "spark.rapids.sql.castFloatToString.enabled": "true",
        "spark.rapids.sql.expression.SortArray": "false"
    })

    def do_it(spark):
        df = gen_df(spark, data_gen, length=100)\
            .groupby('a')\
            .agg(f.collect_set('b').alias("collect_set"))
        # pull out the rdd and schema and create a new dataframe to run SortArray
        # to handle Spark 3.3.0+ optimization that moves SortArray from ProjectExec
        # to ObjectHashAggregateExec
        return spark.createDataFrame(df.rdd, schema=df.schema)\
            .selectExpr("sort_array(collect_set)")

    assert_gpu_and_cpu_are_equal_collect(do_it, conf=conf)
```

File: integration_tests/src/main/python/data_gen.py
Comment (ttnghia) [inline integration_tests/src/main/python/data_gen.py:586]: `min_start` seems to be used only once below. So why don't just:
```
if start is None:
    start = datetime(1, 1, 1, ...)
```

File: integration_tests/src/main/python/data_gen.py
Comment (ttnghia) [inline integration_tests/src/main/python/data_gen.py:592]: Similarly, `max_end` is just used once, so it seems not needed.

Comment (NVnavkumar) [reply integration_tests/src/main/python/data_gen.py:592]: `max_end` is used in the code below if `end` is an instance of `timedelta`, so it's still needed here.

File: integration_tests/src/main/python/data_gen.py
Comment (NVnavkumar) [reply integration_tests/src/main/python/data_gen.py:586]: Fixed.

**Comment by NVnavkumar** [comment]
build

**Comment by NVnavkumar** [comment]
build

**Comment by NVnavkumar** [comment]
build

**Comment by NVnavkumar** [comment]
build

File: integration_tests/src/main/python/hash_aggregate_test.py
Comment (gerashegalov) [inline integration_tests/src/main/python/hash_aggregate_test.py:704]: let us think if this is really the best we can do as it creates quite an overhead.

Comment (NVnavkumar) [reply integration_tests/src/main/python/hash_aggregate_test.py:704]: I've honestly never fully liked this solution anyways (when it was recreating a dataframe and sorting it using the CPU). Thinking about it, it might just be better to do this sort in Python and add functionality to `@ignore_order` for it. Thoughts? It might even be cheaper (for the most part) than doing the whole thing is Spark for <3.3 because that still has to do the sort in CPU Spark anyways.

Comment (NVnavkumar) [reply integration_tests/src/main/python/hash_aggregate_test.py]: Pushed a possible solution to sort the arrays in Python instead of sending it back through Spark. Let me know if this makes more sense.

File: integration_tests/src/main/python/asserts.py
Comment (gerashegalov) [inline integration_tests/src/main/python/asserts.py:223]: please add a high-level comment to explain the algorithm here 

it will read easier if we had another function for a single result that we would apply to from_cpu and from_gpu to eliminate logic duplication

File: integration_tests/src/main/python/hash_aggregate_test.py
Comment (gerashegalov) [inline integration_tests/src/main/python/hash_aggregate_test.py:633]: we do not have many instances of this marker so it looks fine. However, is there anything that is in the way of making the `arrays` parameter implicit.

for tests with this marker can we not just infer the value of `arrays` by looking at the columns with the type `ArrayType`?

**Review by gerashegalov — COMMENTED** [review]
some comments but looks good to me directionally

File: integration_tests/src/main/python/hash_aggregate_test.py
Comment (NVnavkumar) [reply integration_tests/src/main/python/hash_aggregate_test.py:633]: I had actually just thought about this possibility after pushing, will investigate. I wasn't entirely sure about making it implicit, because I wasn't certain about the possibility of array ordering mattering. I could definitely infer the schema though.

Comment (gerashegalov) [reply integration_tests/src/main/python/hash_aggregate_test.py:633]: It is fine to defer it to  a follow-up issue. Probably if we had a convention in test to prefix output columns it could be done quite safely after filtering by type.

**Review by gerashegalov — APPROVED** [review]
LGTM

**Comment by NVnavkumar** [comment]
build

**Comment by NVnavkumar** [comment]
build

**Comment by gerashegalov** [comment]
the [databricks] part is blocked until #10070 is merged

**Comment by NVnavkumar** [comment]
upmerged for Databricks fix.

**Comment by NVnavkumar** [comment]
build

**Comment by NVnavkumar** [comment]
build

**Comment by razajafri** [comment]
build

**Pull request merged**
